### ISSUE TITLE
CrashDaemon can't open coredump file: Permission denied

### ISSUE THREAD

**Issue by tomuta** [SerenityOS/serenity#4698]
I had Browser disappear on me trying to visit the 2nd birthday page. Looks like ProtocolServer asserted. But then no crashdump was displayed. Looks like some permission problem:

```
[NetworkTask(6:6)]: kmalloc(): Adding memory to heap at V0xc0801000, bytes: 1048576
USERSPACE(33) ASSERTION FAILED: m_state == State::InBody
../Libraries/LibHTTP/Job.cpp:210
[ProtocolServer(33:33)]: Terminating ProtocolServer(33) due to signal 6
[FinalizerTask(4:4)]: 0xdeadc0de  Kernel::Processor::switch_context(Kernel::Thread*&, Kernel::Thread*&) +181
0xdeadc0de  Kernel::Scheduler::context_switch(Kernel::Thread*) +155
0xdeadc0de  Kernel::Scheduler::pick_next() +857
0xdeadc0de  Kernel::Processor::check_invoke_scheduler() +125
0xdeadc0de  Kernel::Thread::die_if_needed() +163
0xdeadc0de  syscall_handler +2181
0xdeadc0de  syscall_asm_entry +49
0x015aceca
0x015a4fb9
0x015a58b9
0x0111a8c0
0x0163206c
0x0162cbb6
0x0162d295
0x01622160
0x01622409
0x0166ffdf
0x0167007f
0x1336f11f

[FinalizerTask(4:4)]: Generating coredump for pid: 33
[FinalizerTask(4:4)]: RTC: Year: 2020, month: 12, day: 31, hour: 20, minute: 46, second: 26
CrashDaemon(20): New coredump file: /tmp/coredump/ProtocolServer_33_1609447586
CrashDaemon(20): perror(): open: Permission denied
CrashDaemon(20): Could not open coredump '/tmp/coredump/ProtocolServer_33_1609447586'
```

**Comment by awesomekling** [comment]
Yeah **ProtocolServer** runs as the `protocol` user, while **CrashReporter** runs as `anon`. It makes sense to me that `anon` doesn't get to look into these crashes, but I'm sure we can do something nicer here that still plays by the rules.

cc @bugaevc who often has neat ideas about things like this

**Comment by tomuta** [comment]
Yeah that makes sense. Maybe prompt for root password to elevate privileges? Not sure if we support this...

**Comment by awesomekling** [comment]
That would be pretty slick! cc @linusg 

Perhaps a notification could pop up when there's a crash dump (instead of immediately throwing up the crash reporter)
Activating the notification (this concept doesn't exist yet, but it should!) could then ask you to authenticate if necessary etc.

**Comment by awesomekling** [comment]
Perhaps we could have a privileged service that opens a file on your behalf, makes you authenticate, and if you succeed it passes you an open file descriptor over IPC.

**Comment by alimpfard** [comment]
I just run `CrashReporter /tmp/core<tab><tab>` as root to get it :P

---

Also, that assert is probably because the download is finished (`finish_up()` is called) but the server has more stuff to send, we should just ignore the read event if `m_state == State::Finished` just like cancelled.

**Comment by tomuta** [comment]
The assert is unrelated. The log was just an example of CrashReporter not being able to show the crash dump.

**Comment by bugaevc** [comment]
> Yeah ProtocolServer runs as the protocol user, while CrashReporter runs as anon. It makes sense to me that anon doesn't get to look into these crashes, but I'm sure we can do something nicer here that still plays by the rules.

It makes sense that you cannot look at crash reports of *more* privileged users, but `protocol` is semantically *less* privileged than `anon`. We should sort this out, somehow, without prompting for root access.

`anon` is a member of `protocol` group; perhaps we can alter the coredump permissions in a way that will allow him to access the dump?

**Comment by bcoles** [comment]
> Yeah **ProtocolServer** runs as the `protocol` user, while **CrashReporter** runs as `anon`. It makes sense to me that `anon` doesn't get to look into these crashes, but I'm sure we can do something nicer here that still plays by the rules.

Agreed

> It makes sense that you cannot look at crash reports of _more_ privileged users, but `protocol` is semantically _less_ privileged than `anon`. We should sort this out, somehow, without prompting for root access.
> 
> `anon` is a member of `protocol` group; perhaps we can alter the coredump permissions in a way that will allow him to access the dump?

Perhaps in the short term.

Are we likely to write sensitive information (environment variables, file paths, passwords, other strings, memory ...) to the coredumps in the future?

The concept of "more privileged" is inaccurate when comparing user accounts to shared system services such as `ProtocolServer` and their associated system user accounts. Enforcing core dump access using file permissions based on service group permissions could lead to disclosure of one user's information to another - a "horizontal" elevation of privilege.

Currently, the physical console user with a window session is presumed to be a super-user (ie, `anon`). At least, in a default install when `anon` is in the `wheel` group. The majority of userland services configured in `SystemServer` are also hard coded to use the `anon` account.

In the future, Serenity may support a multi-user environment. While there's no immediate efforts to work towards this goal, this does seem in line with intent. Multi-user is central to UNIX philosophy upon which Serenity is heavily based and inspired.

As such, more than one user could have a window session which would likely require `protocol` group membership for normal operations, and thus also require a `ProtocolServer` session. Enforcing access to coredumps based on group permissions would allow a core dump from one user to be read by another. If `anon` were to cause a crash in `ProtocolServer` while browsing `https://embarassing.local/` they probably wouldn't want this information written out to a core dump file readable by any user in the `protocol` group.

There's also the issue of compromise of any account in the `protocol` group. For whatever reason, the `protocol` user is not a member of the `protocol` group. However, `webcontent` is a member of the `protocol` group. If access to core dumps is enforced by group file system permissions, then in the event that the `webcontent` user account is compromised this account could be used to view all previous `ProtocolServer` core dumps of any user.

**Comment by bugaevc** [comment]
> As such, more than one user could have a window session which would likely require protocol group membership for normal operations, and thus also require a ProtocolServer session.

IIRC, the plan has been, if we ever get to that, to give each user their own "window server UID", their own "protocol server UID", and so on — so `protocol` is in reality more like `protocol_of_anon`
